// ask-smapi-sdk/dist/smapiClientBuilder/AbstractSmapiClientBuilder.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmapiClientBuilder = void 0;
/**
 * Abstract Builder class which should be implemented.
 * @export
 * @class SmapiClientBuilder
 */
class SmapiClientBuilder {
    withCustomUserAgent(userAgent) {
        this.customUserAgent = userAgent;
        return this;
    }
    withAccessTokenConfig(accessTokenConfig) {
        this.accessTokenConfig = accessTokenConfig;
        return this;
    }
    withRefreshTokenConfig(refreshTokenConfig) {
        this.refreshTokenConfig = refreshTokenConfig;
        return this;
    }
    client() {
        throw new Error('client function is not implemented');
    }
    getAuthenticationConfiguration() {
        const authenticationConfiguration = {
            clientId: this.refreshTokenConfig ? this.refreshTokenConfig.clientId : this.accessTokenConfig.clientId,
            clientSecret: this.refreshTokenConfig ? this.refreshTokenConfig.clientSecret : this.accessTokenConfig.clientSecret,
            accessToken: this.accessTokenConfig ? this.accessTokenConfig.accessToken : undefined,
            refreshToken: this.refreshTokenConfig ? this.refreshTokenConfig.refreshToken : undefined
        };
        return authenticationConfiguration;
    }
}
exports.SmapiClientBuilder = SmapiClientBuilder;
